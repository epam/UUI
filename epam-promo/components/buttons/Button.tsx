import * as uui from '@epam/uui';
import { createSkinComponent, devLogger } from '@epam/uui-core';
import { FillStyle } from '../types';
import css from './Button.module.scss';

/** Defines component color. */
type ButtonColor = 'blue' | 'green' | 'red' | 'gray50' | 'gray' | uui.ButtonProps['color'];

type ButtonMods = {
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: FillStyle;
    /**
     * Defines component color.
     * @default 'blue'
     */
    color?: ButtonColor;
    /**
     * Defines component size.
     * @default '36'
     */
    size?: uui.ButtonProps['size'];
};

const mapFill: Record<FillStyle, uui.ButtonProps['fill']> = {
    solid: 'solid',
    white: 'outline',
    light: 'ghost',
    none: 'none',
};

/** Represents the properties of a Button component. */
export type ButtonProps = uui.ButtonCoreProps & ButtonMods;

export const Button = /* @__PURE__ */createSkinComponent<uui.ButtonProps, ButtonProps>(
    uui.Button as any, // TODO: remove it when BaseButton inheritance will be reworked
    (props) => {
        if (__DEV__) {
            devLogger.warnAboutDeprecatedPropValue<ButtonProps, 'color'>({
                component: 'Button',
                propName: 'color',
                propValue: props.color,
                propValueUseInstead: 'gray',
                condition: () => ['gray50'].indexOf(props.color) !== -1,
            });
        }
        return {
            fill: mapFill[props.fill] || mapFill.solid,
        } as any; // TODO: need new helper to rewrite types
    },
    (props) => [
        ['42', '48'].includes(props.size) && css.uppercase,
    ],
);
