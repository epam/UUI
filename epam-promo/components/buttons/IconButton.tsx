import { createSkinComponent } from '@epam/uui-core';
import { EpamPrimaryColor } from '../types';
import * as uui from '@epam/uui';

interface IconButtonMods {
    /**
     * Defines component color.
     * @default 'gray60'
     */
    color?: EpamPrimaryColor | 'gray30' | 'gray50' | 'gray60' | uui.IconButtonProps['color'];
}

/** Represents the properties for the IconButton component. */
export type IconButtonProps = uui.IconButtonCoreProps & IconButtonMods;

export const IconButton = /* @__PURE__ */createSkinComponent<uui.IconButtonProps, IconButtonProps>(
    uui.IconButton,
    (props) => ({ color: props.color ?? 'gray60' }),
);
