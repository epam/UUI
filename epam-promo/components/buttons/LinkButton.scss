@use  '../../assets/styles/index' as *;

.root {
    &:global(.uui-button-box) {
        background-color: transparent;
        color: var(--uui-text);
        fill: var(--uui-text);

        &:global(.-clickable:not(.uui-disabled)) {
            &:hover, &:focus {
                color: var(--uui-text-hover);
                fill: var(--uui-text-hover);
            }

            &:active {
                color: var(--uui-text-active);
                fill: var(--uui-text-active);
            }

            &:focus {
                box-shadow: none;
            }
        }

        &:global(.uui-disabled) {
            color: var(--uui-text-disabled);
            fill: var(--uui-text-disabled);
            background-color: transparent;

            &:hover {
                background-color: transparent;
            }
        }
    }

    :global(.uui-caption) {
        font-family: $font-sans-semibold;
        font-weight: normal;
    }

    // Sizes
    &.size-18 {
        @include link-button-layout(18px);
        min-width: auto;
    }

    &.size-24 {
        @include link-button-layout(24px);
        min-width: auto;
    }

    &.size-30 {
        @include link-button-layout(30px);
        min-width: auto;
    }

    &.size-36 {
        @include link-button-layout(36px);
        min-width: auto;
    }

    &.size-42 {
        @include link-button-layout(42px);
        min-width: auto;

        :global(.uui-caption) {
            font-family: $font-museo-sans;
            text-transform: uppercase;
        }
    }

    &.size-48 {
        @include link-button-layout(48px);
        min-width: auto;

        :global(.uui-caption) {
            font-family: $font-museo-sans;
            text-transform: uppercase;
        }
    }
}
