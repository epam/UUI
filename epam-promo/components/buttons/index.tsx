export * from './Button';
export * from './IconButton';
export * from './LinkButton';
export * from './TabButton';
export * from './VerticalTabButton';
