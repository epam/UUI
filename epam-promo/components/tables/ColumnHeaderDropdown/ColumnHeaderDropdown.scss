.panel {
    @media screen and (max-width: 720px) {
        height: 100vh;
    }
}