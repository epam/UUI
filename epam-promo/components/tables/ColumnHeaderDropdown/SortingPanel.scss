@use  '../../../assets/styles/index' as *;

.filter-sort-button {
    &:global(.-clickable) {
        &:hover {
            background-color: $white;

            .sort-text {
                color: $blue;
            }

            :global(.uui-icon) {
                fill: $blue;
            }
        }

        .active-text {
            color: $blue;
        }
    }

    :global(.uui-icon) {
        margin-left: 12px;
    }

    &:global(.uui-button-box) {
        &:global(.-clickable) {
            &:hover {
                @include button-colors(white, $blue, transparent);
            }
        }
    }
}

.sorting-panel-container {
    padding: 6px 0;
    border-bottom: 1px solid $gray40;
    
    @media screen and (max-width: 720px) {
        padding: 0;
    }
}

.sort-active {
    margin-right: 12px;
}