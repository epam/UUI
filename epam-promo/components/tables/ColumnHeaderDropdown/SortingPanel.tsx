import React, { useCallback } from "react";
import css from "./SortingPanel.scss";
import sortIcon from '@epam/assets/icons/common/table-sort_asc-18.svg';
import sortIconDesc from '@epam/assets/icons/common/table-sort_desc-18.svg';
import { isMobile, SortDirection } from "@epam/uui";
import { FlexCell, FlexRow } from "../../layout";
import { IconButton } from "../../buttons";
import { i18n } from "../../../i18n";
import { Text } from "../../typography";

export interface SortingPanelProps {
    sortDirection: SortDirection;
    onSort(dir: SortDirection): void;
}

const SortingPanelImpl: React.FC<SortingPanelProps> = ({ sortDirection, onSort }) => {
    const sortAsc = useCallback(() => onSort('asc'), [onSort]);
    const sortDesc = useCallback(() => onSort('desc'), [onSort]);
    const size = isMobile() ? "48" : undefined;

    return (
        <FlexCell cx={ css.sortingPanelContainer }>
            <FlexRow size={ size } cx={ css.filterSortButton } spacing="6" onClick={ sortAsc }>
                <IconButton
                    color={ sortDirection === 'asc' ? "blue" : "gray60" }
                    icon={ sortIcon }
                />
                <Text
                    cx={ sortDirection === 'asc' ? css.activeText : css.sortText }
                    color="gray80"
                    fontSize="14"
                    size="24"
                >
                    { i18n.pickerFilterHeader.sortAscending }
                </Text>
            </FlexRow>

            <FlexRow size={ size } cx={ css.filterSortButton } spacing="6" onClick={ sortDesc }>
                <IconButton
                    color={ sortDirection === 'desc' ? "blue" : "gray60" }
                    icon={ sortIconDesc }
                />
                <Text
                    cx={ sortDirection === 'desc' ? css.activeText : css.sortText }
                    color="gray80"
                    fontSize="14"
                    size="24"
                >
                    { i18n.pickerFilterHeader.sortDescending }
                </Text>
            </FlexRow>
        </FlexCell>
    );
};

export const SortingPanel = React.memo(SortingPanelImpl);