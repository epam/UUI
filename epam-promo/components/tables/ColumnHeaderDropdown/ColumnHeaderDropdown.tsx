import React, { Ref, useCallback, useMemo } from 'react';
import css from './ColumnHeaderDropdown.scss';
import { Modifier } from "react-popper";
import { IDropdownToggler, isMobile, mobilePopperModifier } from '@epam/uui';
import { Dropdown, DropdownBodyProps } from '@epam/uui-components';
import { Panel } from '../../layout';
import { MobileDropdownWrapper } from "../../pickers";
import { SortingPanel, SortingPanelProps } from "./SortingPanel";

type ColumnHeaderDropdownProps = SortingPanelProps & {
    isOpen: boolean;
    isSortable: boolean;
    renderTarget: (props: IDropdownToggler & { ref?: Ref<any> }) => React.ReactNode;
    renderFilter?: () => React.ReactNode;
    onOpenChange(nV: boolean): void;
    title: string;
};

const ColumnHeaderDropdownImpl: React.FC<ColumnHeaderDropdownProps> = props => {
    const popperModifiers: Modifier<any>[] = useMemo(() => [
        {
            name: 'offset',
            options: { offset: [0, 1] },
        },
        mobilePopperModifier,
    ], []);

    const style = useMemo(() => ({
        width: isMobile() ? document.documentElement.clientWidth : 350,
    }), []);
    
    const closeDropdown = useCallback(() => props.onOpenChange(false), [props.onOpenChange]);

    return (
        <Dropdown
            renderTarget={ props.renderTarget }
            renderBody={ (dropdownProps: DropdownBodyProps) => (
                <Panel background="white" style={ style } shadow cx={ css.panel }>
                    <MobileDropdownWrapper title={ props.title } close={ closeDropdown }>
                        { props.isSortable && (
                            <SortingPanel
                                sortDirection={ props.sortDirection }
                                onSort={ props.onSort }
                            />
                        ) }
                        { props.renderFilter() }
                    </MobileDropdownWrapper>
                </Panel>
            ) }
            modifiers={ popperModifiers }
            value={ props.isOpen }
            onValueChange={ props.onOpenChange }
        />
    );
};

export const ColumnHeaderDropdown = React.memo(ColumnHeaderDropdownImpl);