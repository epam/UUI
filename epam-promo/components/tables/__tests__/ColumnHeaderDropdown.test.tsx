import React from 'react';
import renderer from 'react-test-renderer';
import { windowMock } from "@epam/test-utils";
import { Button } from '../../buttons';
import { ColumnHeaderDropdown } from '../ColumnHeaderDropdown';

jest.mock('react-dom', () => ({
    createPortal: jest.fn((element, node) => element),
    findDOMNode: jest.fn(),
}));

describe('ColumnHeaderDropdown', () => {
    beforeEach(() => {
        jest.spyOn(window, "window", "get")
            .mockImplementation(() => windowMock);
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });
    
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(<ColumnHeaderDropdown
                renderTarget={ props => <Button caption='Test' { ...props } /> }
                isOpen
                onOpenChange={ jest.fn }
                isSortable
                sortDirection='asc'
                onSort={ jest.fn }
                renderFilter={ () => <div>Picker</div> }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});


