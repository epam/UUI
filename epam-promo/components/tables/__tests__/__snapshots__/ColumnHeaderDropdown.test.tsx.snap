// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnHeaderDropdown should be rendered correctly 1`] = `
Array [
  <button
    className="container uui-button-box uui-enabled -clickable button-vars button-primary root size-36 mode-solid uui-theme-promo"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      Test
    </div>
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className="flipY"
      />
    </div>
  </button>,
  <div
    data-no-focus-lock={true}
  >
    <div
      aria-expanded={true}
      aria-hidden={false}
      className="uui-popper"
      data-placement="bottom-start"
      style={
        Object {
          "left": "0",
          "position": "fixed",
          "top": "0",
          "zIndex": undefined,
        }
      }
    >
      <div
        className="root shadow background-white panel container"
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div
          className="sortingPanelContainer"
          style={
            Object {
              "alignSelf": undefined,
              "flexBasis": "0",
              "flexGrow": undefined,
              "flexShrink": undefined,
              "minWidth": 0,
              "textAlign": undefined,
            }
          }
        >
          <div
            className="itemRoot uui-active container -clickable align-items-center"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="menuitem"
            style={
              Object {
                "columnGap": undefined,
                "rowGap": undefined,
              }
            }
            tabIndex={0}
          >
            <button
              className="container uui-button-box uui-enabled icon-button-default root uui-theme-promo icon-button-color-gray60 icon iconBefore"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <div
                className="container uui-icon uui-enabled"
                style={Object {}}
              >
                <svg
                  className=""
                />
              </div>
            </button>
            <div
              className="caption container"
            >
              Sort Ascending
            </div>
          </div>
          <div
            className="itemRoot container -clickable align-items-center"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="menuitem"
            style={
              Object {
                "columnGap": undefined,
                "rowGap": undefined,
              }
            }
            tabIndex={0}
          >
            <button
              className="container uui-button-box uui-enabled icon-button-default root uui-theme-promo icon-button-color-gray60 icon iconBefore"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <div
                className="container uui-icon uui-enabled"
                style={Object {}}
              >
                <svg
                  className=""
                />
              </div>
            </button>
            <div
              className="caption container"
            >
              Sort Descending
            </div>
          </div>
        </div>
        <div>
          Picker
        </div>
      </div>
    </div>
  </div>,
]
`;
